syntax = "proto3";
package afs;

import "google/protobuf/timestamp.proto";

option go_package = "github.com/SuhasHebbar/CS739-P1/afs";

service AfsService {
	rpc Fetch(Path) returns (stream FileData) {}
	rpc Store(stream StoreData) returns (StoreResult) {}
	rpc Remove(Path) returns  (IOResult) {}
	rpc Create(Path) returns  (IOResult) {}
	rpc Rename(RenameArgs) returns (IOResult) {}
	rpc Makedir(Path) returns  (IOResult) {}
	rpc Removedir(Path) returns  (IOResult) {}
	rpc TestAuth(Path) returns (AuthData) {}
	rpc GetFileStat(Path) returns (StatData) {}
}

message Path {
	optional string name = 1;
}

message FileData {
	optional bytes contents = 1; 
	optional IOResult status = 2;
}

message IOResult {
	optional bool success = 1;
	optional int32 err_code = 2;
}

message StoreData{
	optional Path path = 1;
	optional FileData filedata = 2;
}

message StoreResult{
	optional google.protobuf.Timestamp lmtime = 1;
	optional IOResult status = 2;
}

message AuthData {
	optional google.protobuf.Timestamp lmtime = 1;
	optional IOResult status = 2;
    optional uint32 mode = 3;
}

message StatData {
	optional IOResult status = 1;
	optional DirectoryData dd = 2;
}

message DirectoryData {
	repeated string files = 1;
}

message RenameArgs {
	optional Path oldname = 1;
	optional Path newname = 2;
}

